hashicorp/consul-k8s
# Register mesh-gateways using the endpoints controller.

Changes proposed in this PR:
- Update the mesh-gateway deployment to use the endpoints controller to register the mesh-gateway.
This requires updating the endpoints controller to parse services that are associated with gateway pods and then register those services with Consul as gateway services. Annotations on the gateway pod is used to determine the kind of the gateway as well as any information required to configure the gateway registration.
- Since the endpoints controller manages registration, we no longer need a pre-delete hook to de-register the service. This is performed by the endpoints controller now. Additionally, we don't need a Consul sidecar when metrics are enabled as metrics will be managed by the consul dataplane.
- The init-connect has migrated from acl-init to connect-init as we need to wait for the endpoints controller to register the mesh gateway services before we start the dataplane process. This has allowed significant cleanup of the init container.

How I've tested this PR:
- Added a lot of unit tests to cover gateway registration, parsing the annotations to determine the wan address that the mesh-gateway should get registered with as well as updating the BATS tests.
- Was able to verify that the mesh-gateway acceptance tests run successfully against GKE (they do not run against kind ATM)

How I expect reviewers to test this PR:
- Code review
- Call out places that could potentially need more detailed comments.
- Call out test cases that might be missing.
- Cleaning up for read-abilty as well.
This PR is a little complicated so feel free to pull me into a meeting to ask any questions or understand the changes made here.

Checklist:
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use present tense (e.g. Add support for...)
